gtsfm/frontend/detector_descriptor/sift.py

```diff
@@ -39,28 +39,21 @@ def detect_and_describe(self, image: Image) -> Tuple[Keypoints, np.ndarray]:
             Corr. descriptors, of shape (N, D) where D is the dimension of each descriptor.
         """
 
-        # conert to grayscale
+        # Convert to grayscale.
         gray_image = image_utils.rgb_to_gray_cv(image)
 
-        # Creating OpenCV object
+        # Create OpenCV object.
         opencv_obj = cv.SIFT_create()
 
-        # Run the opencv code
+        # Run the OpenCV code.
         cv_keypoints, descriptors = opencv_obj.detectAndCompute(gray_image.value_array, None)
 
-        # convert to GTSFM's keypoints
+        # Convert to GTSFM's keypoints.
         keypoints = feature_utils.cast_to_gtsfm_keypoints(cv_keypoints)
 
-        # sort the features and descriptors by the score
-        # (need to sort here as we need the sorting order for descriptors)
-        sort_idx = np.argsort(-keypoints.responses)[: self.max_keypoints]
-
-        keypoints = Keypoints(
-            coordinates=keypoints.coordinates[sort_idx],
-            scales=keypoints.scales[sort_idx],
-            responses=keypoints.responses[sort_idx],
-        )
-
-        descriptors = descriptors[sort_idx]
+        # Filter features.
+        if image.mask is not None:
+            keypoints, descriptors = self.filter_by_mask(image.mask, keypoints, descriptors)
```

Comment:
you can use opencv itself for the mask.

Reply:
I'm not aware of an OpenCV function that masks keypoints. Can you point me to the function you're referring to?

Reply:
Snippet: "sift.detect() function finds the keypoint in the images. You can pass a mask if you want to search only a part of image" from https://docs.opencv.org/3.4/da/df5/tutorial_py_sift_intro.html.

Hence I prefer the detection algorithm to handle the masks rather than us doing it. The function you wrote to filter out detections in the masked region can go to utils, and we can call it in the `detect_and_describe` function for algorithms like `Superpoint` which do not provide a direct API to apply a mask. Let's not make it a post-processing step for every algorithm.

Reply:
Yes, I'm aware of the one for features implemented by OpenCV (e.g., SIFT); I thought you were referring to a general keypoint masking function.

You're right about the OpenCV masking for the _handcrafted_ features (e.g., ORB, SIFT) being different than just masking the output, as they mask the image at each level in their scale pyramid. Thanks for bringing this up.

Reply:
Now passing mask to `detectAndCompute`